### PR TITLE
Add context information to thrown errors

### PR DESCRIPTION
**Pull request by LaurensRietveld**
The PR following our discussion on #109

### CODE REVIEW

**Comment by LaurensRietveld** [comment]
I added some context checks to the unit tests. Doing it everywhere seemed a bit overkill, so I added it to a select few.

File: lib/N3Parser.js [hunk 1]
```diff
@@ -775,8 +775,11 @@ N3Parser.prototype = {
   // ### `_error` emits an error message through the callback
   _error: function (message, token) {
     var err = new Error(message + ' on line ' + token.line + '.');
-    err.token = token;
-    err.line = token.line;
+    err.context = {
+      token: token,
+      line: token.line,
+      previousToken: this._lexer._prevToken,
```
Comment (LaurensRietveld) [inline lib/N3Parser.js]: I'm not sure whether this line might become an issue. The token is passed as argument, where the previous token is fetched from the lexer directly. Is there a scenario where both are not 'in sync'?

Comment (LaurensRietveld) [inline lib/N3Parser.js]: ps. afaik this doesnt seem to be an issue in the unit tests I updated

**Comment by RubenVerborgh** [comment]
Thanks, will take this up soon!

File: lib/N3Lexer.js [hunk 2]
```diff
@@ -319,9 +319,10 @@ N3Lexer.prototype = {
       }
 
       // Emit the parsed token
-      callback(null, { line: line, type: type, value: value, prefix: prefix });
+      var token = { line: line, type: type, value: value, prefix: prefix };
+      callback(null, token);
       this._prevTokenType = type;
-
+      this._prevToken = token;
```
Comment (RubenVerborgh) [inline lib/N3Lexer.js]: Should check the performance impact (should be minimal).

File: lib/N3Parser.js [hunk 3]
```diff
@@ -774,7 +774,13 @@ N3Parser.prototype = {
 
   // ### `_error` emits an error message through the callback
   _error: function (message, token) {
-    this._callback(new Error(message + ' on line ' + token.line + '.'));
+    var err = new Error(message + ' on line ' + token.line + '.');
+    err.context = {
+      token: token,
+      line: token.line,
+      previousToken: this._lexer._prevToken,
```
Comment (RubenVerborgh) [inline lib/N3Parser.js]: Alternatively, the parser might store it. Would avoid accessing a private variable. (Also, a little strange that the lexer exposes `_prevToken` but not `_token`.)

**Comment by RubenVerborgh** [comment]
Added a couple of minor things to think about.

File: lib/N3Lexer.js (hunk 2 above)
Comment (LaurensRietveld) [inline lib/N3Lexer.js]: did some eyeballing on the N3Lexer-perf script (ran on an ntriple file with 4million statements). Memory footprint is the same. User time _might_ have decreased, but with around 1 or 2% at most

File: lib/N3Parser.js (hunk 3 above)
Comment (LaurensRietveld) [inline lib/N3Parser.js]: You mean by adding the prevToken as a parameter in the tokenize callback? (https://github.com/OpenTriply/N3.js/blob/8ccffc54904a3f5a5b669f8f6f434c22b2ac1746/lib/N3Parser.js#L897)
I'm comfortable with just accessing the private variable tbh

**Comment by LaurensRietveld** [comment]
Added some responses. Using it now btw in some of my libs. Works great!

**Comment by RubenVerborgh** [comment]
Thanks, merged as 4601c2a8d5e54a1d9f4d53b3ca33f4f87dfe9c16.

**Pull request closed without merging**